theano/tensor/nnet/bn.py

```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+    if gamma.ndim != ndim or beta.ndim != ndim:
+        raise ValueError("gamma and beta must be of the same dimensionality "
+                         "as inputs; got %d and %d instead of %d" %
+                         (gamma.ndim, beta.ndim, ndim))
+    if mean.ndim != ndim or var.ndim != ndim:
+        raise ValueError("mean and var must be of the same dimensionality "
+                         "as inputs; got %d and %d instead of %d" %
+                         (mean.ndim, var.ndim, ndim))
+    if epsilon < 1e-5:
+        raise ValueError("epsilon must be at least 1e-5, got %f" % epsilon)
+
+    batchnorm_op = AbstractBatchNormInference(mode=mode)
+    return batchnorm_op(inputs, gamma, beta, mean, var, epsilon=epsilon)
+
+
+class AbstractBatchNorm(Op):
+    """
+    Abstract Op for Batch Normalization.
+
+    Parameters
+    ----------
+    mode : {'per-activation', 'spatial'}
+        Whether to normalize per activation (in this mode, bias and scale
+        tensor dimensions are 1xCxHxW) or share normalization factors across
+        spatial dimensions (in this mode, bias and scale tensor dimensions
+        are 1xCx1x1).
+    epsilon
+        Epsilon value used in the batch normalization formula. Minimum allowed
+        value is 1e-5 (imposed by cuDNN).
+    """
+
+    __props__ = ('mode',)
+
+    def __init__(self, mode='per-activation'):
+        assert (mode in ('per-activation', 'spatial'))
+        self.mode = mode
+
+    def infer_shape(self, node, shape):
+        return [shape[0], shape[1], shape[1]]
+
+    def make_node(self, x, scale, bias, epsilon=1e-4):
+        assert x.ndim == scale.ndim == bias.ndim
+        if not isinstance(epsilon, theano.Variable):
+            epsilon = as_tensor_variable(epsilon)
+        return Apply(self, [x, scale, bias, epsilon], [x.type(), scale.type(), scale.type()])
+
+    def grad(self, inputs, grads):
+        x, scale, bias, epsilon = inputs
+        dy = grads[0]
+        _, x_mean, x_invstd = self.make_node(x, scale, bias, epsilon).outputs
+        return AbstractBatchNormGrad(self.mode)(x, dy, scale, x_mean,
+                                                x_invstd, epsilon) + [theano.gradient.DisconnectedType()()]
+
+    def connection_pattern(self, node):
+        # Specificy that epsilon is not connected to outputs.
+        return [[True, True, True], [True, True, True], [True, True, True],
+                [False, False, False]]
+
+    def perform(self, node, inputs, output_storage):
+        x, scale, bias, epsilon = inputs
+
+        if self.mode == 'per-activation':
+            axes = (0,)
+        elif self.mode == 'spatial':
+            axes = (0,) + tuple(range(2, x.ndim))
+        else:
+            raise ValueError('invalid mode: %s', str(self.mode))
+
+        mean = x.mean(axes, keepdims=True)
+        stdinv = 1.0 / numpy.sqrt(x.var(axes, keepdims=True) + epsilon)
+        out = (x - mean) * scale * stdinv + bias
```

Comment:
This should have `(scale * stdinv)` in brackets for improved performance (multiply the smaller tensors first).